### DISCUSSION TITLE
Questions on MaterialGrid

### DISCUSSION THREAD

**Discussion by mawc2019** [NanoComp/meep#2240] [Q&A] [answered]
According to the Meep [documents](https://meep.readthedocs.io/en/latest/Python_User_Interface/#materialgrid), the grid points of `MaterialGrid` are defined at the corners of the voxels. Therefore, if we want to discretize a `MaterialGrid` region with a specified `resolution`, it seems that the numbers of grid points should be calculated like
`Nx,Ny = int(resolution*width)+1,int(resolution*height)+1` instead of `Nx,Ny = int(resolution*width),int(resolution*height)`.
In the latter case, the spacing between adjacent grid points is slightly different from `1/resolution`, which means the actual resolution of the `MaterialGrid` is slightly different from the given `resolution`. But in the [tutorials](https://github.com/NanoComp/meep/blob/master/python/examples/adjoint_optimization/03-Filtered_Waveguide_Bend.ipynb) and [built-in tests](https://github.com/NanoComp/meep/blob/b00f2d4ce20e683275a4895c8dc8318677a296d1/python/tests/test_adjoint_solver.py), `Nx` and `Ny` are computed as the latter expression. Is it a convention in Meep that resolution is defined according to the number of points rather than the number of intervals like `1/𐤃x`?

Another question is related to periodic boundary conditions. If a `MaterialGrid` region extends across the periodic direction, e.g., the _x_ direction, the grid points at the smallest-_x_ boundary should always have the same `weights` as the grid points at the largest-_x_ boundary because of periodicity, but why does `MaterialGrid` allow them to be different, or do I have misunderstanding here?

**Comment by stevengj** [comment, marked as answer]
Yes, the `+1` form is correct.   I think we recently changed some of the tutorials and/or tests to use this `+ 1`, in fact?

**Reply by smartalecH** [reply]
> I think we recently changed some of the tutorials and/or tests to use this + 1, in fact?

We actually changed it *back* to *not* using +1. See #2016

**Reply by stevengj** [reply]
Why was that?

**Reply by smartalecH** [reply]
As discussed today we should change it back -- sorry!

**Comment by stevengj** [comment]
Hmm, I can't find this; @oskooi, do you remember this change?

**Comment by mawc2019** [comment]
In addition, would it be better to replace expressions like `int(Lx * resolution)` and `int(Ly * resolution)` with `int(round(Lx * resolution))` and `int(round(Ly * resolution))` in [cylindrical_filter](https://github.com/NanoComp/meep/blob/b00f2d4ce20e683275a4895c8dc8318677a296d1/python/adjoint/filters.py#L89), [conic_filter](https://github.com/NanoComp/meep/blob/b00f2d4ce20e683275a4895c8dc8318677a296d1/python/adjoint/filters.py#L135), and [gaussian_filter](https://github.com/NanoComp/meep/blob/b00f2d4ce20e683275a4895c8dc8318677a296d1/python/adjoint/filters.py#L181),@smartalecH ? The former expression sometimes produce unexpected results. For example, suppose `resolution=40` and `Lx=1+1/resolution+1/resolution`. People may expect `int(Lx * resolution)` to be 42, but the result is 41, which can be demonstrated as follows.
```
resolution = 40
Lx = 1+1/resolution+1/resolution
print("int(Lx*resolution) = ",int(Lx*resolution))
print("int(round(Lx*resolution)) = ",int(round(Lx*resolution)))
```
The output is
```
int(Lx*resolution) =  41
int(round(Lx*resolution)) =  42
```

**Comment by mawc2019** [comment]
According to the Meep [documents](https://meep.readthedocs.io/en/latest/Python_User_Interface/#materialgrid), the grid points are defined at the corners of the voxels, and the `weights` array is used to define a linear interpolation from `medium1` to `medium2`. If so, even if the `weights` array becomes binarized, the refractive indices represented by `weights` may not be binarized in the regions between adjacent grid points with different `weights` values, i.e., 0 and 1. Do I have misunderstanding here?

**Comment by stevengj** [comment]
Remember that the projection happens *after* the linear interpolation, so it is still binary almost everywhere for β=∞. 

However, when it is then discretized onto the Yee grid, there is subpixel smoothing.
